hydra-main/src/main/java/com/addthis/hydra/job/spawn/balancer/SpawnBalancer.java

```diff
@@ -1404,15 +1404,33 @@ private JobTaskItemByHostMap generateTaskCountByHost(List<HostState> hosts, Iter
     }
 
     private List<JobTaskMoveAssignment> pushTasksOntoHost(HostState host, Iterable<HostState> heavyHosts) {
+        int moveLimit = config.getTasksMovedFullRebalance();
         MoveAssignmentList moveAssignments = new MoveAssignmentList(spawn, taskSizer);
+        List<JobTaskMoveAssignment> rv = purgeMisplacedTasks(host, moveLimit);
+
+        List<HostState> lightHostList = new ArrayList<>();
+        lightHostList.add(host);
+
+        double byteLimitFactor = 1;
+
         for (HostState heavyHost : heavyHosts) {
-            double byteLimitFactor =
-                    1 - ((double) moveAssignments.getBytesUsed() / config.getBytesMovedFullRebalance());
-            moveAssignments.addAll(pushTasksOffHost(heavyHost, Collections.singletonList(host), true, byteLimitFactor,
-                                                    config.getTasksMovedFullRebalance(), true));
+            if (moveAssignments.size() >= moveLimit || !canReceiveNewTasks(host) || moveAssignments.getBytesUsed() >= config.getBytesMovedFullRebalance()) {
+                break;
+            }
+            List<JobTaskMoveAssignment> assignments = pushTasksOffHost(heavyHost, lightHostList,true, byteLimitFactor, 1, true);
```

Comment:
`pushTasksOffHost` calls `markRecentlyReplicatedTo` to mark assigned hosts as recently-replicated-to. But because of the same-job check below, the assignment may be ignored, resulting in incorrect marking of the host.